File: backend/internal/handlers/agenda.go
package handlers

import (
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"

	"github.com/inovar/backend/internal/middleware"
	"github.com/inovar/backend/internal/models"
)

// GetAgenda returns agenda entries
func (h *Handler) GetAgenda(c *fiber.Ctx) error {
	userID := middleware.GetUserID(c)
	role := middleware.GetUserRole(c)
	companyID := middleware.GetCompanyID(c)

	// Query params
	start := c.Query("start")
	end := c.Query("end")
	technicianID := c.Query("technicianId")

	var agenda []models.Agenda
	query := h.DB.Preload("Solicitacao")

	// Filter by user/company
	switch role {
	case models.RoleTecnico:
		query = query.Where("user_id = ?", userID)
	case models.RolePrestador:
		// All technicians from company
		query = query.Where("user_id IN (SELECT id FROM users WHERE company_id = ?)", companyID)
	}

	// Filter by technician if provided
	if technicianID != "" {
		query = query.Where("user_id = ?", technicianID)
	}

	// Filter by date range
	if start != "" && end != "" {
		startTime, _ := time.Parse("2006-01-02", start)
		endTime, _ := time.Parse("2006-01-02", end)
		query = query.Where("scheduled_at >= ? AND scheduled_at <= ?", startTime, endTime.Add(24*time.Hour))
	}

	query.Order("scheduled_at ASC").Find(&agenda)

	return Success(c, agenda)
}

// CreateAgendaRequest represents agenda entry creation
type CreateAgendaRequest struct {
	UserID        string `json:"userId,omitempty"`
	SolicitacaoID string `json:"solicitacaoId"`
	Title         string `json:"title"`
	ScheduledAt   string `json:"scheduledAt"`
	Duration      int    `json:"duration"`
	Notes         string `json:"notes,omitempty"`
}

// CreateAgendaEntry creates a new agenda entry
func (h *Handler) CreateAgendaEntry(c *fiber.Ctx) error {
	userID := middleware.GetUserID(c)
	role := middleware.GetUserRole(c)

	var req CreateAgendaRequest
	if err := c.BodyParser(&req); err != nil {
		return BadRequest(c, "Dados inválidos")
	}

	// For technicians, use their own ID
	targetUserID := req.UserID
	if role == models.RoleTecnico || targetUserID == "" {
		targetUserID = userID
	}

	scheduledAt, err := time.Parse(time.RFC3339, req.ScheduledAt)
	if err != nil {
		return BadRequest(c, "Data inválida")
	}

	agenda := models.Agenda{
		ID:            uuid.New().String(),
		UserID:        targetUserID,
		SolicitacaoID: req.SolicitacaoID,
		Title:         req.Title,
		ScheduledAt:   scheduledAt,
		Duration:      req.Duration,
		Notes:         req.Notes,
		CreatedAt:     time.Now(),
	}

	h.DB.Create(&agenda)

	// Update solicitacao status and scheduled date
	h.DB.Model(&models.Solicitacao{}).Where("id = ?", req.SolicitacaoID).Updates(map[string]interface{}{
		"status":       models.StatusAgendada,
		"scheduled_at": scheduledAt,
	})

	h.Hub.Broadcast("agenda:created", agenda)

	return Created(c, agenda)
}

// UpdateAgendaEntry updates an agenda entry
func (h *Handler) UpdateAgendaEntry(c *fiber.Ctx) error {
	id := c.Params("id")

	var agenda models.Agenda
	if err := h.DB.First(&agenda, "id = ?", id).Error; err != nil {
		return NotFound(c, "Agendamento não encontrado")
	}

	var req CreateAgendaRequest
	if err := c.BodyParser(&req); err != nil {
		return BadRequest(c, "Dados inválidos")
	}

	if req.ScheduledAt != "" {
		scheduledAt, _ := time.Parse(time.RFC3339, req.ScheduledAt)
		agenda.ScheduledAt = scheduledAt

		// Update solicitacao scheduled date
		h.DB.Model(&models.Solicitacao{}).Where("id = ?", agenda.SolicitacaoID).Update("scheduled_at", scheduledAt)
	}

	if req.Duration > 0 {
		agenda.Duration = req.Duration
	}
	if req.Notes != "" {
		agenda.Notes = req.Notes
	}
	if req.Title != "" {
		agenda.Title = req.Title
	}

	agenda.UpdatedAt = time.Now()
	h.DB.Save(&agenda)

	h.Hub.Broadcast("agenda:updated", agenda)

	return Success(c, agenda)
}

// DeleteAgendaEntry deletes an agenda entry
func (h *Handler) DeleteAgendaEntry(c *fiber.Ctx) error {
	id := c.Params("id")

	var agenda models.Agenda
	if err := h.DB.First(&agenda, "id = ?", id).Error; err != nil {
		return NotFound(c, "Agendamento não encontrado")
	}

	// Start transaction
	tx := h.DB.Begin()

	// 1. Clear scheduled date in solicitacao
	if err := tx.Model(&models.Solicitacao{}).Where("id = ?", agenda.SolicitacaoID).Update("scheduled_at", nil).Error; err != nil {
		tx.Rollback()
		return ServerError(c, err)
	}

	// 2. Delete agenda entry
	if err := tx.Delete(&agenda).Error; err != nil {
		tx.Rollback()
		return ServerError(c, err)
	}

	tx.Commit()

	h.Hub.Broadcast("agenda:deleted", fiber.Map{"id": id})

	return Success(c, fiber.Map{"message": "Agendamento removido"})
}

File: infra/pyservice/routes/agenda.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List, Optional
import uuid
from datetime import datetime

from database import get_db
from models import Agenda, User, Solicitacao
from pydantic import BaseModel

router = APIRouter()

class AgendaCreate(BaseModel):
    user_id: str
    solicitacao_id: str
    title: str
    scheduled_at: datetime
    duration: int = 60
    notes: Optional[str] = None

@router.get("/")
def list_agenda(
    user_id: Optional[str] = None,
    company_id: Optional[str] = None,
    start: Optional[datetime] = None,
    end: Optional[datetime] = None,
    db: Session = Depends(get_db)
):
    query = db.query(Agenda)

    if user_id:
        query = query.filter(Agenda.user_id == user_id)
    elif company_id:
        query = query.join(User).filter(User.company_id == company_id)

    if start:
        query = query.filter(Agenda.scheduled_at >= start)
    if end:
        query = query.filter(Agenda.scheduled_at <= end)

    return {"data": query.order_by(Agenda.scheduled_at.asc()).all()}

@router.post("/")
def create_agenda(req: AgendaCreate, db: Session = Depends(get_db)):
    entry = Agenda(
        id=str(uuid.uuid4()),
        user_id=req.user_id,
        solicitacao_id=req.solicitacao_id,
        title=req.title,
        scheduled_at=req.scheduled_at,
        duration=req.duration,
        notes=req.notes
    )
    db.add(entry)

    # Update solicitacao
    sol = db.query(Solicitacao).filter(Solicitacao.id == req.solicitacao_id).first()
    if sol:
        sol.status = "AGENDADA"
        sol.scheduled_at = req.scheduled_at

    db.commit()
    db.refresh(entry)
    return {"data": entry}

@router.put("/{id}")
def update_agenda(id: str, req: dict, db: Session = Depends(get_db)):
    entry = db.query(Agenda).filter(Agenda.id == id).first()
    if not entry:
        raise HTTPException(404, "Not found")

    for field, value in req.items():
        if hasattr(entry, field):
            if field == "scheduled_at" and value:
                if isinstance(value, str):
                    value = datetime.fromisoformat(value.replace("Z", "+00:00"))
                # Update sol too
                sol = db.query(Solicitacao).filter(Solicitacao.id == entry.solicitacao_id).first()
                if sol:
                    sol.scheduled_at = value
            setattr(entry, field, value)

    db.commit()
    db.refresh(entry)
    return {"data": entry}

@router.delete("/{id}")
def delete_agenda(id: str, db: Session = Depends(get_db)):
    entry = db.query(Agenda).filter(Agenda.id == id).first()
    if not entry:
        raise HTTPException(404, "Not found")

    # Clear sol
    sol = db.query(Solicitacao).filter(Solicitacao.id == entry.solicitacao_id).first()
    if sol:
        sol.scheduled_at = None

    db.delete(entry)
    db.commit()
    return {"message": "Deleted"}
